sourcecred/sourcecred
# Add Update CLI for magical instance updates

# Description
Breaking changes are a PAIN for instance maintainers because of all the json hand editing n whatnot. We can release versions of the sourcecred package but a big chunk of our product exists in the instance forks, where the instance maintainers are responsible for making updates. Now, with this PR, instance maintainers can use a new CLI to auto-magically update their instance using code that we write.

I included arg support for power users, but ultimately, I rely on the updater.txt state added in this PR and the fact that updaters are going to be released with their version (meaning that you can't update beyond your current version). I decided against doing any package.json reading/writing because I want this to work well with dependabot.

The 0.9.0 updater included is for breaking changes made in https://github.com/sourcecred/sourcecred/pull/3064

# The new UX
1. [unchanged] Dependabot creates a version bump PR when a sourcecred version is released
2. [unchanged] The Instance Maintainer checks out the branch locally to test and make necessary changes
3. The Instance Maintainer runs `yarn sourcecred update`
4. If there are any updaters available, the Instance Maintainer sees files changed, including the updater.txt, and commits the changes along with the dependabot commit.

# Test plan

1. build and run `scdev update` and verify new updater.txt and changed discord config
2. run it again and verify no change
3. Add `["asdf", async () => {}]` to registry, build and run it again and verify it runs the new updater and updates the txt
1. Delete updater.txt and reset discord config and registry
2. build and run `scdev update 0.9.0` and verify new updater.txt and changed discord config
2. build and run `scdev update 123` and verify error is thrown
2. Delete updater.txt, reset registry, and delete config/plugins
3. build and run `scdev update` and verify the updater completes gracefully

Before:
```
{
  "channelWeightConfig": {
    "defaultWeight": 1,
    "weights": {}
  },
  "guildId": "678348980639498428",
  "includeNsfwChannels": false,
  "propsChannels": [],
  "reactionWeights": {},
  "roleWeightConfig": {
    "defaultWeight": 1,
    "weights": {}
  }
}
```
After:
```
[
  {
    "channelWeightConfig": {
      "defaultWeight": 1,
      "weights": {
      }
    },
    "guildId": "678348980639498428",
    "includeNsfwChannels": false,
    "propsChannels": [
    ],
    "reactionWeightConfig": {
      "applyAveraging": false,
      "defaultWeight": 1,
      "weights": {
      }
    },
    "roleWeightConfig": {
      "defaultWeight": 1,
      "weights": {
      }
    }
  }
]
```

Example console output:
```
thena@thenas-mbp template-instance % scdev update
Found and successfully ran upgrader for 0.9.0

Remember to commit and push changes to data/updater.txt
thena@thenas-mbp template-instance % scdev update     
No updaters found since 0.9.0.
```
